GuruExample.java

```diff
@@ -0,0 +1,49 @@
+/* An example of code that should get flagged by CodeGuru */
+import java.util.HashMap;
+import java.io.InputStreamReader;
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.IOException;
+
+class Whatsit {
+    public int doit(int val) {
+        return 12345;
+    }
+}
+
+class GuruExample {
+    public static void main(String args[]) throws IOException {
+        int len = Integer.parseInt(args[0]);
+        char[] thing = new char[len];
+
+        for (int i=0; i<thing.length; i++) {
+            thing[i] = 'x';
+        }
+
+        System.out.println("Did a thing");
+
+        HashMap<String, Whatsit> thang = new HashMap<String, Whatsit>();
+        foo(thang);
+
+        spang();
+    }
+
+    public static synchronized void foo(HashMap<String, Whatsit> thing) {
+        Whatsit whatsit = thing.get("banana");
+        if (whatsit != null) {
+            int result = whatsit.doit(123);
+            thing.put("banana", whatsit);
+        }
+    }
+
+    public static void spang() throws IOException {
+        Process p = Runtime.getRuntime().exec("date");
+        BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. The resource is referenced in statements at the following lines: 41, 44. The resource closure statement is at line: 45. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by BufferedReader.readLine. Either a) close *in* in a try-finally block or b) close the resource by declaring *in* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).